### PR TITLE
Disallow auto_element_wise for in-place and fix some in-place gradients

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79056
* __->__ #85634

Fixes https://github.com/pytorch/pytorch/issues/85535

Also fixes the backward and forward gradients of `nn.functional.threshold`. The issue was that in-place gradients weren't tested because the in-place variants were not properly registered to the OpInfo.

Perhaps an alternative to this to make auto_element_wise smart enough to actually handle the in-places cases (we have 4 cases total now where we manually copy_ after doing auto_element_wise), but that requires a few more changes.

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2020,8 +2020,8 @@
   result: auto_element_wise
 
 - name: threshold_(Tensor(a!) self, Scalar threshold, Scalar value) -> Tensor(a!)
-  self: threshold_backward(grad, result, threshold)
-  result: auto_element_wise
+  self: threshold_backward(grad, self, threshold)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:2023]: That's not ok I think.
If you do `threshold(self, a, b)` with `b > a`, this will be wrong!

**Review by albanD — COMMENTED** [review]
The change to the backward formula sounds wrong.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:2023]: Oh I think that was happening before this change, but not anymore? 
We want to do `grad_input = self > threshold ? grad_output : zero` right?

Comment (albanD) [reply tools/autograd/derivatives.yaml]: 1. We should have such a case in the OpInfo samples
2. We want to check self yes IIRC (use OpInfo testing to see which one is correct to be sure)
3. Changing this can have significant impact on memory usage patterns. This is the inplace relu op that is used everywhere. Using `self` here means that it will be cloned: extra buffer!

Comment (soulitzer) [reply tools/autograd/derivatives.yaml]: > 1. We should have such a case in the OpInfo samples
> 2. We want to check self yes IIRC (use OpInfo testing to see which one is correct to be sure)

We don't have an explicit check because all values are chosen at random, but that means it will be checked with high probability.

> 3. Changing this can have significant impact on memory usage patterns. This is the inplace relu op that is used everywhere. Using self here means that it will be cloned: extra buffer!

Okay I checked and relu doesn't actually use threshold (and since a == b for relu, so we can still be better off saving result instead of self). For threshold though I guess that now it makes less sense to use in-place (at least when gradients are required).

**Review by albanD — APPROVED** [review]
SGTM
